bluez/plugins/jolla-wakelock.c

```diff
@@ -0,0 +1,293 @@
+	if (r < 0) {
+		error("Failed to release real lock for lock: %p, name: %s",
+			lock, lock->name);
+		return r;
+	}
+
+	real_lock_acquired = FALSE;
+	DBG("Real wakelock released for lock: %p, name %s", lock, lock->name);
+
+done:
+	DBG("lock: %p, name %s released", lock, lock->name);
+	return 0;
+}
+
+static int jolla_wakelock_create(const gchar *name, struct wakelock **lock)
+{
+	DBG("name: %s", name);
+
+	*lock = g_new0(struct wakelock, 1);
+	(*lock)->name = g_strdup(name);
+	(*lock)->acquisitions = 0;
+	(*lock)->stale = FALSE;
+
+	locks = g_slist_prepend(locks, *lock);
+
+	DBG("lock: %p created", *lock);
+	return 0;
+}
+
+static int jolla_wakelock_free(struct wakelock *lock)
+{
+	DBG("lock: %p, name: %s, acquisitions: %u, stale: %s",
+		lock, lock->name, lock->acquisitions,
+		lock->stale ? "yes" : "no");
+
+	if (lock->acquisitions) { /* Force release */
+		warn("Freeing unreleased lock: %p", lock);
+		lock->acquisitions = 1;
+		jolla_wakelock_release(lock);
+	}
+
+	locks = g_slist_remove(locks, lock);
+	g_free(lock->name);
+	g_free(lock);
+
+	DBG("lock: %p freed", lock);
+	return 0;
+}
+
+static void jolla_wakelock_exit(void)
+{
+	DBG("");
+
+	if (!initialized)
+		return;
+
+	/* May happen due to plugin exit ordering */
+	if (locks) {
+		DBG("Marking remaining wakelocks as stale");
+		while (locks) {
+			struct wakelock *lock = locks->data;
+			if (lock->acquisitions) { /* Force release */
+				lock->acquisitions = 1;
+				jolla_wakelock_release(lock);
+			}
+			lock->stale = TRUE;
+			locks = g_slist_remove(locks, lock);
+		}
+	}
+
+	wakelock_plugin_unregister();
+	initialized = FALSE;
+	DBG("Wakelocks uninitialized");
+
+	/* Sanity: force release just in case our internal state got messed up */
+	sysfs_write(UNLOCK_FILE, LOCK_NAME, LOCK_NAME_LEN);
+}
+
+static GKeyFile *load_config(const char *file)
+{
+	GError *err = NULL;
+	GKeyFile *keyfile;
+
+	keyfile = g_key_file_new();
+
+	g_key_file_set_list_separator(keyfile, ',');
+
+	if (!g_key_file_load_from_file(keyfile, file, 0, &err)) {
+		error("Parsing %s failed: %s", file, err->message);
+		g_error_free(err);
+		g_key_file_free(keyfile);
+		return NULL;
+	}
+
+	return keyfile;
+}
+
+static int jolla_wakelock_init(void)
+{
+	static struct wakelock_table table = {
+		.create = jolla_wakelock_create,
+		.free = jolla_wakelock_free,
+		.acquire = jolla_wakelock_acquire,
+		.release = jolla_wakelock_release
+	};
+	gboolean use_wakelocks = FALSE;
+	GKeyFile *config;
+	int r;
+
+	DBG("");
+
+	if (initialized)
+		return -EALREADY;
+
+	config = load_config(CONFIGDIR "/jolla.conf");
+	if (config) {
+		use_wakelocks = g_key_file_get_boolean(config,
+							"General",
+							"Wakelocks",
+							NULL);
+		g_key_file_free(config);
+	}
+
+	if (!use_wakelocks) /* No need to initialize or uninitialize */
+		return 0;
+
+	/* Sanity: release any left over wakelock from previous processes
+	   (e.g. after a crash-restart) */
+	sysfs_write(UNLOCK_FILE, LOCK_NAME, LOCK_NAME_LEN);
```

Comment:
If there is no lock to cleanup, sysfs_write() can hit EINVAL on write & possibly spam the journal - might be nice to avoid that.